File: Java/201709/20170904/Main.java
import java.util.HashSet;
import java.util.Scanner;
import java.util.Set;

public class Main {
	static int[][] graph1 = new int[1001][1001];
	static int[][] graph2 = new int[1001][1001];
	static int N, M;
	static int[] visit1 = new int[1001];
	static int[] visit2 = new int[1001];
	static int[] visit = new int[1001];
	static Set<Integer> record = new HashSet<Integer>();
	public static void main(String[] args) {
		Scanner scanner = new Scanner(System.in);
		N = scanner.nextInt();
		M = scanner.nextInt();
		for (int i = 0; i < M; i++) {
			int s = scanner.nextInt();
			int t = scanner.nextInt();
			graph1[s][t] = 1;
			graph2[t][s] = 1;
		}
		
		for (int i = 1; i <= N; i++) {
			init(1, 1, 1);
			dfs(i, graph1, visit1);
//			System.out.println();
			if (check())
				record.add(i);
			else {
				init(0, 1, 1);
				dfs(i, graph2, visit2);
//				System.out.println();
				if (check())
					record.add(i);
			}
		}
		
		System.out.println(record.size());
	}
	
	public static void dfs(int v, int[][] graph, int[] visited) {
//		System.out.print(v + " ");
		visited[v] = 1;
		for (int i = 1; i <= N; i++)
			if (graph[v][i] == 1 && visited[i] == 0)
				dfs(i, graph, visited);
	}
	
	public static void init(int a, int b, int c) {
		for (int i = 1; i <= N; i++) {
			if (a == 1)
				visit1[i] = 0;
			if (b == 1)
				visit2[i] = 0;
			if (c == 1)
				visit[i] = 0;
		}
	}
	
	public static boolean check() {
		for (int i = 1; i <= N; i++)
		{
			visit[i] = visit1[i] + visit2[i];
		}
		for (int i = 1; i <= N; i++)
			if (visit[i] == 0)
				return false;
		return true;
	}
}

File: C++/201709/20170904/main.cpp
#include <stdio.h>
#include <string.h>

using namespace std;

int graph1[1001][1001];
int graph2[1001][1001];
int N, M;
int visit1[1001], visit2[1001], visit[1001];

void dfs(int v, int visited[1001], int graph[1001][1001]);
bool check();
void init(int a, int b, int c);

int main()
{
    scanf("%d %d", &N, &M);
    for (int i = 0; i < M; i++)
    {
        int s, t;
        scanf("%d %d", &s, &t);
        graph1[s][t] = 1;
        graph2[t][s] = 1;
    }

    int count = 0;

    for (int i = 1; i <= N; i++)
    {
        init(1, 1, 1);
        dfs(i, visit1, graph1);
        if (check())
            count ++;
        else
        {
            init(0, 1, 1);
            dfs(i, visit2, graph2);
            if (check())
                count++;
        }
    }

    printf("%d", count);

    return 0;
}

void dfs(int v, int visited[1001], int graph[1001][1001])
{
    visited[v] = 1;
    for (int i = 1; i <= N; i++)
        if (visited[i] == 0 && graph[v][i] == 1)
            dfs(i, visited, graph);
}

bool check()
{
    for (int i = 1; i <= N; i++)
        visit[i] = visit1[i] + visit2[i];
    for (int i = 1; i <= N; i++)
        if (visit[i] == 0)
            return false;
    return true;
}

void init(int a, int b, int c)
{
    if (a == 1)
        memset(visit1, 0, sizeof(visit1));
    if (b == 1)
        memset(visit2, 0, sizeof(visit2));
    if (c == 1)
        memset(visit, 0, sizeof(visit));
}
